client/network/src/request_responses.rs

```diff
@@ -0,0 +1,864 @@
+		&mut self,
+		peer_id: PeerId,
+		connection: ConnectionId,
+		(p_name, event): <Self::ProtocolsHandler as ProtocolsHandler>::OutEvent,
+	) {
+		if let Some((proto, _)) = self.protocols.get_mut(&*p_name) {
+			return proto.inject_event(peer_id, connection, event)
+		}
+
+		log::warn!(target: "sub-libp2p",
+			"inject_node_event: no request-response instance registered for protocol {:?}",
+			p_name)
+	}
+
+	fn inject_new_external_addr(&mut self, addr: &Multiaddr) {
+		for (p, _) in self.protocols.values_mut() {
+			NetworkBehaviour::inject_new_external_addr(p, addr)
+		}
+	}
+
+	fn inject_expired_listen_addr(&mut self, addr: &Multiaddr) {
+		for (p, _) in self.protocols.values_mut() {
+			NetworkBehaviour::inject_expired_listen_addr(p, addr)
+		}
+	}
+
+	fn inject_dial_failure(&mut self, peer_id: &PeerId) {
+		for (p, _) in self.protocols.values_mut() {
+			NetworkBehaviour::inject_dial_failure(p, peer_id)
+		}
+	}
+
+	fn inject_new_listen_addr(&mut self, addr: &Multiaddr) {
+		for (p, _) in self.protocols.values_mut() {
+			NetworkBehaviour::inject_new_listen_addr(p, addr)
+		}
+	}
+
+	fn inject_listener_error(&mut self, id: ListenerId, err: &(dyn std::error::Error + 'static)) {
+		for (p, _) in self.protocols.values_mut() {
+			NetworkBehaviour::inject_listener_error(p, id, err)
+		}
+	}
+
+	fn inject_listener_closed(&mut self, id: ListenerId, reason: Result<(), &io::Error>) {
+		for (p, _) in self.protocols.values_mut() {
+			NetworkBehaviour::inject_listener_closed(p, id, reason)
+		}
+	}
+
+	fn poll(
+		&mut self,
+		cx: &mut Context,
+		params: &mut impl PollParameters,
+	) -> Poll<
+		NetworkBehaviourAction<
+			<Self::ProtocolsHandler as ProtocolsHandler>::InEvent,
+			Self::OutEvent,
+		>,
+	> {
+		'poll_all: loop {
+			// Poll to see if any response is ready to be sent back.
+			while let Poll::Ready(Some(result)) = self.pending_responses.poll_next_unpin(cx) {
+				match result {
+					RequestProcessingOutcome::Response {
+						protocol, inner_channel, response
+					} => {
+						if let Some((protocol, _)) = self.protocols.get_mut(&*protocol) {
+							protocol.send_response(inner_channel, response);
+						}
+					}
+					RequestProcessingOutcome::Busy { peer, protocol } => {
+						let out = Event::InboundRequest {
+							peer,
+							protocol,
+							result: Err(ResponseFailure::Busy),
+						};
+						return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					}
+				}
+			}
+
+			// Poll request-responses protocols.
+			for (protocol, (behaviour, resp_builder)) in &mut self.protocols {
+				while let Poll::Ready(ev) = behaviour.poll(cx, params) {
+					let ev = match ev {
+						// Main events we are interested in.
+						NetworkBehaviourAction::GenerateEvent(ev) => ev,
+
+						// Other events generated by the underlying behaviour are transparently
+						// passed through.
+						NetworkBehaviourAction::DialAddress { address } => {
+							log::error!("The request-reponse isn't supposed to start dialing peers");
```

Comment:
```suggestion
							log::error!("The request-response isn't supposed to start dialing peers");
```
This is the context of `DialAddr` btw. There is also `DialPeer` just below where you probably want to log something similar.